conwayste/src/uimanager.rs

```diff
@@ -0,0 +1,165 @@
+            Rc::clone(&font),
+            "Toggle FullScreen".to_owned(),
+            Rect::new(10.0, 210.0, 20.0, 20.0),
+        ));
+
+
+        let mut layer_mainmenu = Layer::new(WidgetID::MainMenuLayer1);
+        let mut layer_ingame = Layer::new(WidgetID::InGameLayer1);
+
+        // Create a new pane, and add two test buttons to it. Actions do not really matter for now, WIP
+        let mut pane = Box::new(Pane::new(WidgetID::MainMenuPane1, Rect::new_i32(20, 20, 300, 250)));
+        let mut pane_button = Box::new(Button::new(ctx, WidgetID::MainMenuPane1ButtonYes,
+            UIAction::ScreenTransition(Screen::ServerList),
+            WidgetID::MainMenuPane1ButtonYesLabel,
+            Rc::clone(&font),
+            "ServerList".to_owned()
+        ));
+        pane_button.set_size(Rect::new(10.0, 10.0, 180.0, 50.0));
+        pane.add(pane_button);
+
+        let mut pane_button = Box::new(Button::new(ctx, WidgetID::MainMenuPane1ButtonNo,
+            UIAction::ScreenTransition(Screen::InRoom),
+            WidgetID::MainMenuPane1ButtonNoLabel,
+            Rc::clone(&font),
+            "InRoom".to_owned()
+        ));
+        pane_button.set_size(Rect::new(10.0, 70.0, 180.0, 50.0));
+        pane.add(pane_button);
+
+        let mut pane_button = Box::new(Button::new(ctx, WidgetID::MainMenuTestButton,
+            UIAction::ScreenTransition(Screen::Run),
+            WidgetID::MainMenuTestButtonLabel,
+            Rc::clone(&font),
+            "StartGame".to_owned()
+        ));
+        pane_button.set_size(Rect::new(10.0, 130.0, 180.0, 50.0));
+        pane.add(pane_button);
+
+        pane.add(checkbox);
+
+        layer_mainmenu.add(pane);
+        layer_ingame.add(chatpane);
+
+        ui_layers.insert(Screen::Menu, vec![layer_mainmenu]);
+        ui_layers.insert(Screen::Run, vec![layer_ingame]);
+
+        UIManager {
+            layers: ui_layers,
+        }
+    }
+
+
+    /// Returns a reference to the layer's currently focused text field
+    pub fn focused_textfield_mut(&mut self, screen: Screen) -> Option<&mut TextField> {
+        if let Some(layer) = self.get_top_layer_from_screen(screen) {
+            if let Some(id) = layer.focused_widget {
+                let widget = layer.get_widget_mut(id);
+                return widget.downcast_mut::<TextField>();
+            }
+        }
+        None
+    }
+
+    /// Get the current screen's top most layer
+    pub fn get_top_layer_from_screen(&mut self, screen: Screen) -> Option<&mut Layer> {
+        if let Some(vec_layer) = self.layers.get_mut(&screen) {
+            return vec_layer.last_mut();
+        }
+        None
+    }
+
+    pub fn get_screen_layers(&mut self, screen:Screen) -> Option<&mut Vec<Layer>> {
+        self.layers.get_mut(&screen)
+    }
+
+    pub fn textfield_from_id(&mut self, screen: Screen, id: WidgetID) -> Option<&mut TextField> {
+        if let Some(layer) = self.get_top_layer_from_screen(screen) {
```

Comment:
Any particular reason why you're not iterating through all the layers to find one with that ID?

Reply:
A user shouldn't be allowed to interact with a widget if it's at a lower layer, so there's no point in iterating over all layers. We just want to check the top-most layer as that is what it shown the user.